flent/runners.py

```diff
@@ -845,6 +846,95 @@ def find_binary(cls, host, interval, length, ip_version, local_bind=None,
         raise RuntimeError("No suitable Iperf binary found.")
 
 
+class ParseException(Exception):
+    pass
+
+
+class SsRunner(ProcessRunner):
+    """Runner for iterated `ss -t -i -p`. Depends on same partitial output
+    separationa and time stamping as TcRunner."""
+
+    ip_v4_addr_sub_re = "(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])"
+    ip_v6_addr_sub_re = "(?:(?:[A-F0-9]{1,4}:){7}[A-F0-9]{1,4}|(?=(?:[A-F0-9]{0,4}:){0,7}[A-F0-9]{0,4}(?![:.\w]))(([0-9A-F]{1,4}:){1,7}|:)((:[0-9A-F]{1,4}){1,7}|:))(:\d+)+"
+
+    time_re = re.compile(r"^Time: (?P<timestamp>\d+\.\d+)", re.MULTILINE)
+    cwnd_re = re.compile(r"cwnd:(?P<cwnd>\d+)", re.MULTILINE)
+    rtt_re = re.compile(r"rtt:(?P<rtt>\d+.\d+/\d+.\d+)", re.MULTILINE)
+    pid_re = re.compile(r"pid=(?P<pid>\d+)", re.MULTILINE)
+    ports_ipv4_re = re.compile(r"" + ip_v4_addr_sub_re +
+                               ":(?P<src_p>\d+)\s+?" + ip_v4_addr_sub_re + ":(?P<dst_p>\d+)")
+    ports_ipv6_re = re.compile(r"" + "(?P<src_p>" + ip_v6_addr_sub_re + ")" + "\s+" +
+                               "(?P<dst_p>" + ip_v6_addr_sub_re + ")", re.IGNORECASE)
+
+    src_p = []
+    dst_p = []
+    pid = []
+
+    def parse_part(self, part):
+        timestamp = self.time_re.search(part)
+        cwnd = self.cwnd_re.search(part)
+        rtt = self.rtt_re.search(part)
+        pid = self.pid_re.search(part)
+
+        f_ports = self.ports_ipv4_re.search(part)
+        if None is f_ports:
+            f_ports = self.ports_ipv6_re.search(part)
+
+        if None in [timestamp, cwnd, rtt, pid, f_ports]:
+            raise ParseException()
+
+        timestamp = float(timestamp.group('timestamp'))
+        cwnd = int(cwnd.group('cwnd'))
+        rtt, rtt_var = [float(x) for x in rtt.group('rtt').split('/')]
+        pid = int(pid.group('pid'))
+        self.src_p.append(f_ports.group('src_p').split(":")[-1])
+        self.dst_p.append(f_ports.group('dst_p').split(":")[-1])
+
+        val_res = {'cwnd': cwnd, 'rtt': rtt, 'rtt_var': rtt_var}
+
+        return timestamp, val_res, pid
+
+    def parse(self, output):
+        results = {}
+        parts = output.split("\n---\n")
+        for part in parts:
+            try:
+                t, res_dict, pid = self.parse_part(part)
+                for k, v in res_dict.items():
+                    if k not in results:
+                        results[k] = [[t, v]]
+                    else:
+                        results[k].append([t, v])
+
+                self.pid.append(pid)
+                raw_values = copy.deepcopy(results)
```

Comment:
Raw values should basically be what is currently res_dict, but with the timestamp as the 't' key. No need to copy (and see suggestion above about just having parse_part return the raw_values dict).